File: features/user/chat_common/src/main/java/com/tokopedia/chat_common/view/adapter/viewholder/MessageViewHolder.java
package com.tokopedia.chat_common.view.adapter.viewholder;

import android.content.Context;
import android.text.TextUtils;
import android.view.View;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;

import androidx.annotation.LayoutRes;

import com.tokopedia.abstraction.common.utils.view.MethodChecker;
import com.tokopedia.chat_common.R;
import com.tokopedia.chat_common.data.MessageViewModel;
import com.tokopedia.chat_common.util.ChatLinkHandlerMovementMethod;
import com.tokopedia.chat_common.view.adapter.viewholder.listener.ChatLinkHandlerListener;

/**
 * @author by nisie on 5/16/18.
 */
public class MessageViewHolder extends BaseChatViewHolder<MessageViewModel> {

    private static final String ROLE_USER = "User";

    private Context context;
    protected TextView message;
    private ImageView chatStatus;
    private View chatBalloon;
    private TextView name;
    private TextView label;
    private TextView dot;

    @LayoutRes
    public static final int LAYOUT = R.layout.layout_message_chat;


    public MessageViewHolder(View itemView, ChatLinkHandlerListener listener) {
        super(itemView);
        this.context = itemView.getContext();
        message = itemView.findViewById(getMessageId());
        message.setMovementMethod(new ChatLinkHandlerMovementMethod(listener));
        chatStatus = itemView.findViewById(getChatStatusId());
        name = itemView.findViewById(getNameId());
        label = itemView.findViewById(getLabelId());
        dot = itemView.findViewById(getDotId());
        chatBalloon = itemView.findViewById(getMainId());
        hour = itemView.findViewById(getHourId());
        date = itemView.findViewById(getDateId());

    }

    @Override
    protected int getHourId() {
        return R.id.hour;
    }

    @Override
    protected int getDateId() {
        return R.id.date;
    }

    protected int getChatStatusId() {
       return R.id.chat_status;
    }

    protected int getMessageId() {
        return R.id.message;
    }
    protected int getNameId() {
        return R.id.name;
    }
    protected int getLabelId() {
        return R.id.label;
    }
    protected int getDotId() {
        return R.id.dot;
    }
    protected int getMainId() {
        return R.id.main;
    }

    @Override
    public void bind(MessageViewModel element) {
        super.bind(element);
        message.setText(MethodChecker.fromHtml(element.getMessage()));
        setupChatBubbleAlignment(chatBalloon, element);
        setRole(element);
    }

    private void setupChatBubbleAlignment(View chatBalloon, MessageViewModel element) {
        if (element.isSender()) {
            setChatRight(chatBalloon);
            setReadStatus(element);
        } else {
            setChatLeft(chatBalloon);
        }
    }

    protected void setChatLeft(View chatBalloon) {
        chatBalloon.setBackground(context.getResources().getDrawable(R.drawable
                .left_bubble));
        setAlignParent(RelativeLayout.ALIGN_PARENT_LEFT, chatBalloon);
        setAlignParent(RelativeLayout.ALIGN_PARENT_LEFT, hour);
        message.setTextColor(MethodChecker.getColor(context, com.tokopedia.unifyprinciples.R.color.Unify_N700_68));
        message.setLinkTextColor(MethodChecker.getColor(context, com.tokopedia.unifyprinciples.R.color.Unify_N700_68));
        chatStatus.setVisibility(View.GONE);
    }

    private void setAlignParent(int alignment, View view) {
        RelativeLayout.LayoutParams params = (RelativeLayout.LayoutParams) view.getLayoutParams();
        params.addRule(RelativeLayout.ALIGN_PARENT_LEFT, 0);
        params.addRule(RelativeLayout.ALIGN_PARENT_RIGHT, 0);
        params.addRule(alignment);
        view.setLayoutParams(params);
    }

    protected void setChatRight(View chatBalloon) {
        chatBalloon.setBackground(context.getResources().getDrawable(R.drawable
                .right_bubble));
        setAlignParent(RelativeLayout.ALIGN_PARENT_RIGHT, chatBalloon);
        setAlignParent(RelativeLayout.ALIGN_PARENT_RIGHT, hour);
        message.setTextColor(MethodChecker.getColor(context, com.tokopedia.unifyprinciples.R.color.Unify_N0));
        message.setLinkTextColor(MethodChecker.getColor(context, com.tokopedia.unifyprinciples.R.color.Unify_N0));
        chatStatus.setVisibility(View.VISIBLE);
    }

    private void setRole(MessageViewModel element) {
        if (!TextUtils.isEmpty(element.getFromRole())
                && !element.getFromRole().toLowerCase().equals(ROLE_USER.toLowerCase())
                && element.isSender()
                && !element.isDummy()
                && element.isShowRole()) {
            name.setText(element.getFrom());
            label.setText(element.getFromRole());
            name.setVisibility(View.VISIBLE);
            dot.setVisibility(View.VISIBLE);
            label.setVisibility(View.VISIBLE);

        } else {
            name.setVisibility(View.GONE);
            label.setVisibility(View.GONE);
            dot.setVisibility(View.GONE);
        }
    }

    private void setReadStatus(MessageViewModel element) {
        int imageResource;
        if (element.isShowTime()) {
            chatStatus.setVisibility(View.VISIBLE);
            if (element.isRead()) {
                imageResource = R.drawable.ic_chat_read;
            } else {
                imageResource = R.drawable.ic_chat_unread;
            }

            if (element.isDummy()) {
                imageResource = R.drawable.ic_chat_pending;
            }
            chatStatus.setImageDrawable(MethodChecker.getDrawable(chatStatus.getContext(),imageResource));
        } else {
            chatStatus.setVisibility(View.GONE);
        }
    }
}

File: features/communication/chat_common/src/main/java/com/tokopedia/chat_common/view/adapter/viewholder/MessageViewHolder.kt
package com.tokopedia.chat_common.view.adapter.viewholder

import android.content.Context
import android.text.TextUtils
import android.view.View
import android.widget.ImageView
import android.widget.RelativeLayout
import android.widget.TextView
import androidx.annotation.LayoutRes
import com.tokopedia.abstraction.common.utils.view.MethodChecker
import com.tokopedia.chat_common.R
import com.tokopedia.chat_common.data.MessageUiModel
import com.tokopedia.chat_common.util.ChatLinkHandlerMovementMethod
import com.tokopedia.chat_common.view.adapter.viewholder.listener.ChatLinkHandlerListener
import java.util.Locale

/**
 * @author by nisie on 5/16/18.
 */
open class MessageViewHolder(
    itemView: View, listener: ChatLinkHandlerListener?
) : BaseChatViewHolder<MessageUiModel>(itemView) {

    private val context: Context = itemView.context

    @JvmField
    protected var message: TextView
    private val chatStatus: ImageView? by lazy {
        itemView.findViewById(chatStatusId)
    }
    private val chatBalloon: View? by lazy {
        itemView.findViewById(mainId)
    }
    private val name: TextView? by lazy {
        itemView.findViewById(nameId)
    }
    private val label: TextView? by lazy {
        itemView.findViewById(labelId)
    }
    private val dot: TextView? by lazy {
        itemView.findViewById(dotId)
    }

    override val hourId: Int
        get() = R.id.hour
    override val dateId: Int
        get() = R.id.date
    override val chatStatusId: Int
        get() = R.id.chat_status
    protected open val messageId: Int
        get() = R.id.message
    protected open val nameId: Int
        get() = R.id.name
    protected open val labelId: Int
        get() = R.id.label
    protected open val dotId: Int
        get() = R.id.dot
    protected open val mainId: Int
        get() = R.id.main

    override fun bind(uiModel: MessageUiModel) {
        super.bind(uiModel)
        uiModel.let {
            message.text = MethodChecker.fromHtml(it.message)
            chatBalloon?.let { view ->
                setupChatBubbleAlignment(view, it)
            }

            setRole(it)
        }
    }

    private fun setupChatBubbleAlignment(chatBalloon: View, element: MessageUiModel) {
        if (element.isSender) {
            setChatRight(chatBalloon)
            setReadStatus(element)
        } else {
            setChatLeft(chatBalloon)
        }
    }

    protected open fun setChatLeft(chatBalloon: View) {
        chatBalloon.background = context.resources.getDrawable(R.drawable.left_bubble)
        setAlignParent(RelativeLayout.ALIGN_PARENT_LEFT, chatBalloon)
        hour?.let {
            setAlignParent(RelativeLayout.ALIGN_PARENT_LEFT, it)
        }
        message.setTextColor(
            MethodChecker.getColor(
                context,
                com.tokopedia.unifyprinciples.R.color.Unify_NN950_68
            )
        )
        message.setLinkTextColor(
            MethodChecker.getColor(
                context,
                com.tokopedia.unifyprinciples.R.color.Unify_NN950_68
            )
        )
        chatStatus?.visibility = View.GONE
    }

    private fun setAlignParent(alignment: Int, view: View) {
        val params = view.layoutParams as RelativeLayout.LayoutParams
        params.addRule(RelativeLayout.ALIGN_PARENT_LEFT, 0)
        params.addRule(RelativeLayout.ALIGN_PARENT_RIGHT, 0)
        params.addRule(alignment)
        view.layoutParams = params
    }

    protected open fun setChatRight(chatBalloon: View) {
        chatBalloon.background =
            context.resources.getDrawable(R.drawable.right_bubble)
        setAlignParent(RelativeLayout.ALIGN_PARENT_RIGHT, chatBalloon)
        hour?.let {
            setAlignParent(RelativeLayout.ALIGN_PARENT_RIGHT, it)
        }

        message.setTextColor(
            MethodChecker.getColor(
                context,
                com.tokopedia.unifyprinciples.R.color.Unify_NN0
            )
        )
        message.setLinkTextColor(
            MethodChecker.getColor(
                context,
                com.tokopedia.unifyprinciples.R.color.Unify_NN0
            )
        )
        chatStatus?.visibility = View.VISIBLE
    }

    private fun setRole(element: MessageUiModel) {
        if (!TextUtils.isEmpty(element.fromRole)
            && element.fromRole.lowercase(Locale.getDefault()) != ROLE_USER.lowercase(Locale.getDefault())
            && element.isSender
            && !element.isDummy
            && element.isShowRole
        ) {
            name?.text = element.from
            label?.text = element.fromRole
            name?.visibility = View.VISIBLE
            dot?.visibility = View.VISIBLE
            label?.visibility = View.VISIBLE
        } else {
            name?.visibility = View.GONE
            label?.visibility = View.GONE
            dot?.visibility = View.GONE
        }
    }

    private fun setReadStatus(element: MessageUiModel) {
        var imageResource: Int
        if (element.isShowTime) {
            chatStatus?.visibility = View.VISIBLE
            imageResource = if (element.isRead) {
                R.drawable.ic_chat_read
            } else {
                R.drawable.ic_chat_unread
            }
            if (element.isDummy) {
                imageResource = R.drawable.ic_chat_pending
            }
            chatStatus?.let {
                it.setImageDrawable(
                    MethodChecker.getDrawable(
                        it.context,
                        imageResource
                    )
                )
            }

        } else {
            chatStatus?.visibility = View.GONE
        }
    }

    companion object {
        private const val ROLE_USER = "User"

        @LayoutRes
        val LAYOUT = R.layout.layout_message_chat
    }

    init {
        message = itemView.findViewById(messageId)
        message.movementMethod = ChatLinkHandlerMovementMethod(listener)
    }
}